### ISSUE TITLE
[Android] 3DSecure confirmation screen closed if app is put in background

### ISSUE THREAD

**Issue by konstantinmv** [stripe/stripe-react-native#355]
Hi all. 
We currently have a bug happening on Android only. 
After using `confirmPaymentMethod` on a payment, the 3DS screen is brought up. If instead of confirming / denying the payment on this screen we go into background and then bring the app back to active again, 3DS screen is just dismissed, so the app is stuck. 
Obviously, at that moment Stripe payment is pending, but there's no way to bring up the payment screen again. 
Even if there was, then the process would be repeated - if the user has to go to the bank's app in order to confirm the payment or whatever, the problem would come back and always be present.

Note: this does not happen on iOs. 

**To Reproduce**
Steps to reproduce the behaviour:
1. Boot Android React Native app. 
2. Confirm payment with a 3DS card. 
3. When the `hooks.stripe.com` screen with the confirmation gets brought up, go into background. 
4. Wait 2 seconds, open the app again. 
5. There you have a problem. 

**Expected behavior**
The expected behaviour would be the same as in iOs, for this screen not to be dismissed until the user does some action upon it. 

**Desktop (please complete the following information):**
OS: MacOs Big Sur 11.2.3
Browser: Chrome
Version: 91.0.4472.77 (Official Build) (x86_64)

**Smartphone (please complete the following information):**
Device: Honor 9 Lite
Android version: 9

**Additional context**
Stripe version: 0.1.4
React native version: 0.63.4

Thanks in advance!

**Comment by thorsten-stripe** [comment]
@michelleb-stripe can you look into this?

**Comment by vipu-sa** [comment]
React Native Android - The previous state of stripe payment screen not getting When app reopens click of app icon after minimize.
when I am performing 3d secure payment using stripe payment screen.When I minimize the application in between this process and, reopen it by clicking on the app icon at that time I am not getting the previous state of stripe payment screen specific for android.

**Comment by aanalmehta-sa** [comment]
We are facing the same issue. By when we can expect this issue to be resolved?
We are having a major bug because of this.

**Comment by michelleb-stripe** [comment]
@aanalmehta-sa and @vipu-sa and @konstantinmv 

In your manifest file you will have a single activity that has an intent filter of category: `android.intent.category.LAUNCHER` similar to:
```
      <activity
        android:name=".MainActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize|uiMode"
        android:launchMode="singleTask"
        android:windowSoftInputMode="adjustResize">
        <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
```

Can you post the launchMode set on this activity?  I believe this might be an issue with a launchMode = singleTask.  If you use the defaultLaunch mode it might work.   (See related issue: https://github.com/stripe/stripe-android/issues/3290).

**Comment by arekkubaczkowski** [comment]
closing due to inactivity, @aanalmehta-sa @vipu-sa @konstantinmv please reopen if above solution didn't work for you

**Comment by cristeahub** [comment]
> closing due to inactivity, @aanalmehta-sa @vipu-sa @konstantinmv please reopen if above solution didn't work for you

@arekkubaczkowski I want to draw attention to https://github.com/stripe/stripe-android/issues/3290 also being closed pointing here, while the issue is still real for users like myself. Please reconsider reopening again.

> Can you post the launchMode set on this activity? I believe this might be an issue with a launchMode = singleTask. If you use the defaultLaunch mode it might work. (See related issue: stripe/stripe-android#3290).

@michelleb-stripe I have this issue and I am using `launchMode="singleTask"`. My current React Native app will not work if I change my launchMode, I am unable to use `defaultLaunch`.

**Comment by aaronSig** [comment]
I'm having the same issue. 

To recreate:
- Initiate payment
- Leave the app to check for a 3D secure SMS.
- Re-open the app by clicking on the app icon on the home screen

The payment errors out with "Failed to retrieve a PaymentSheetResult".

However, if you instead resume the app from the Android Task Switcher everything works as expected. 


My manifest looks like:
```
<activity
    android:name=".MainActivity"
    android:label="@string/app_name"
    android:configChanges="keyboard|keyboardHidden|orientation|screenSize|uiMode"
    android:launchMode="singleTask"
    android:windowSoftInputMode="adjustResize">
    
    <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
    </intent-filter>

    <intent-filter>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="myapp" />
    </intent-filter>	
</activity>
```

I believe you're right that changing the singleTask launch mode will fix but this breaks deep linking for our RN Android app.

Would there be another work around?

**Comment by michelleb-stripe** [comment]
@acomley-stripe @trag-stripe Can you help engage with the expo team on this particular issue?  I am happy to help jump in a conversation if needed.

**Comment by tautvilas** [comment]
I am experiencing same issue. If launchMode is not singleTask resuming app from app icon works. However we can not use other launch mode with react-native because it might introduce other issues. Also when returning back to the app from the app switcher works, but if app is resumed from the app icon the 3Ds confirmation view is no more.

**Comment by MagicMirouff** [comment]
Same issue here. 3DS process is killed when we resume app from home with appIcon ...
Any solution ?

**Comment by cfeeney5** [comment]
Is there any update on this bug? Its causing a lot of issue for customers

**Comment by Djauron** [comment]
any update ?

**Comment by charliecruzan-stripe** [comment]
Looking into this, looks like an issue with the underlying android SDK, but could be wrong

**Comment by viraj-vermut** [comment]
Is there any update on this bug? We are facing the same issue for android users.

**Comment by ottob** [comment]
Same problem here on Android with the following config (managed flow expo app):

expo: 46.0.17
react-native: 0.69.6
@stripe/stripe-react-native: 0.13.1

**Comment by nicolasdevienne** [comment]
Same issue with `presentPaymentSheet`

**Comment by ottob** [comment]
Hi, are you making any progress on this?

**Comment by charliecruzan-stripe** [comment]
The native version of the 3DS2 auth screen is on the roadmap and is the long term fix for this issue, but we're still looking into proper workarounds in the meantime

**Comment by huextrat** [comment]
Hi @charliecruzan-stripe, any news regarding this issue ? It has always been there for 2 years and no solution is provided.

This is an issue affecting the 3DS confirmation on Android and causes unintended cases about payment, it's a bit annoying.

A workaround would be appreciable in next release until we get a real fix.

**Comment by charliecruzan-stripe** [comment]
Unfortunately we're still waiting on a fix in the `stripe-android` repo, I agree that this is pretty annoying and has been for a while. Let me check and see what the status of this fix is

**Comment by ottob** [comment]
Friendly ping :)

**Comment by tgensol** [comment]
@charliecruzan-stripe any news ? (Can you give us the issue on stripe-android, thanks )

**Comment by tgensol** [comment]
@charliecruzan-stripe We made a quick fixed it on our own. Do you have any news ?

**Comment by ottob** [comment]
> We made a quick fixed it on our own.

@tgensol can you please share how you fixed it in your project?

**Comment by gvestit** [comment]
Hi,

We have an web app but with the users using Android faced this issue. Something that we can do?

**Comment by huextrat** [comment]
@charliecruzan-stripe any update on this issue ?

**Comment by stianjensen** [comment]
Hi guys! I think this is an issue that cannot really be solved by the Stripe library itself, as it has to do with the overall architecture of the Android app using it. Luckily, the fix should be quite straight-forward in a simple react-native app!

I found it when researching a similar issue in another part of our app where we display a Chrome custom tab in order to perform login authentication:
https://github.com/proyecto26/react-native-inappbrowser/issues/213#issuecomment-878915862

Basically, instead of just having one MainActivity where deeplink support requires you to give the MainActivity a launchMode of "singleTask", you split the application into two activites. One that is responsible for launching the app from the android launcher, and one that is then started by that 'launcher activity' and retains the singleTask launchmode.

I think the only thing Stripe can really do about this is to include this information in their documentation somehow, for users who need both deeplink support (for instance via Branch) and the option to use 3D secure via webviews.

**Comment by huextrat** [comment]
Hi @stianjensen, thank you for this reply!

We've implemented it with a few adaptations for React Native while keeping deep links working properly, and it seems to work very well.

**Comment by zdnk** [comment]
Is there a way to achieve this in Expo managed workflow? or will you provide it as part of Stripe plugin for Expo?

**Comment by zdnk** [comment]
[Here](https://github.com/expo/expo/issues/8072) they have been trying to solve similar issue. They forced the web browser to NOT create a new task. Is this what might fix the issue? How can we force it?

**Comment by zdnk** [comment]
This is still a big issue for us using expo, anyone has a solution? We have many users failing the payment because of this issue.

**Comment by fabrii** [comment]
Hi @charliecruzan-stripe 

I ran into a case where deep links are not received in MainActivity when using a LaunchActivity.

When opening a deep link from a cold start (app fully closed), Android delivers the intent with the deep link URI to LaunchActivity, but MainActivity receives a new intent with null data.

As a result, Linking.getInitialURL() (or equivalent) in JS returns null.

To fix this, I had to manually forward the deep link intent from LaunchActivity to MainActivity, like this:

```kotlin
if (!app.isActivityInBackStack(MainActivity::class.java)) {
    val originalIntent = intent
    val newIntent = Intent(this, MainActivity::class.java).apply {
        action = originalIntent?.action
        data = originalIntent?.data
        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
    }
    startActivity(newIntent)
}
```

```java
if (!app.isActivityInBackStack(MainActivity.class)) {
    Intent originalIntent = getIntent();

    Intent newIntent = new Intent(this, MainActivity.class);
    newIntent.setAction(originalIntent.getAction());
    newIntent.setData(originalIntent.getData());
    newIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);

    startActivity(newIntent);
}
```
